File: examples/buffer_reuse.rs
use bstack::BStack;
use std::io;

fn main() -> io::Result<()> {
    let stack = BStack::open("buffer_reuse_example.bstack")?;

    // Push some data
    stack.push(b"First message\n")?; // 14 bytes
    stack.push(b"Second message\n")?; // 15 bytes
    stack.push(b"Third message\n")?; // 14 bytes

    println!("Stack length: {} bytes", stack.len()?);

    // Demonstrate get_into for zero-copy reading
    let mut buf = vec![0u8; 14]; // Buffer for "First message\n"
    stack.get_into(0, &mut buf)?;
    println!("Read with get_into: {:?}", String::from_utf8_lossy(&buf));

    // Demonstrate pop_into for zero-copy popping (pops last 14 bytes: "Third message\n")
    let mut pop_buf = vec![0u8; 14];
    stack.pop_into(&mut pop_buf)?;
    println!(
        "Popped with pop_into: {:?}",
        String::from_utf8_lossy(&pop_buf)
    );

    println!("Stack length after pop: {} bytes", stack.len()?);

    // Peek remaining data
    let remaining = stack.peek(0)?;
    println!("Remaining data: {:?}", String::from_utf8_lossy(&remaining));

    Ok(())
}

File: examples/buffer_reuse.c
/* Demonstrates that bstack_get and bstack_pop write into caller-supplied
 * buffers — no extra allocation is needed for reads. */

#include "../c/bstack.h"

#include <stdio.h>
#include <stdlib.h>

int main(void)
{
    bstack_t *stack = bstack_open("buffer_reuse_example.bstack");
    if (!stack) { perror("bstack_open"); return 1; }

    bstack_push(stack, (const uint8_t *)"First message\n",  14, NULL);
    bstack_push(stack, (const uint8_t *)"Second message\n", 15, NULL);
    bstack_push(stack, (const uint8_t *)"Third message\n",  14, NULL);

    uint64_t len;
    bstack_len(stack, &len);
    printf("Stack length: %llu bytes\n", (unsigned long long)len);

    /* bstack_get fills a caller-supplied buffer — no allocation */
    uint8_t buf[14];
    bstack_get(stack, 0, 14, buf);
    printf("Read with bstack_get: \"%.*s\"\n", 14, buf);

    /* bstack_pop removes bytes and writes them into a caller-supplied buffer */
    uint8_t pop_buf[14];
    size_t written;
    bstack_pop(stack, 14, pop_buf, &written);   /* removes "Third message\n" */
    printf("Popped with bstack_pop: \"%.*s\"\n", (int)written, pop_buf);

    bstack_len(stack, &len);
    printf("Stack length after pop: %llu bytes\n", (unsigned long long)len);

    uint8_t *remaining = malloc((size_t)len);
    bstack_peek(stack, 0, remaining, &written);
    printf("Remaining data: \"%.*s\"\n", (int)written, remaining);
    free(remaining);

    bstack_close(stack);
    return 0;
}
